api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java

```diff
@@ -461,6 +492,11 @@ public void unregisterListeners(Plugin plugin)
      */
     public Collection<Map.Entry<String, Command>> getCommands()
     {
-        return Collections.unmodifiableCollection( commandMap.entrySet() );
+        Map<String, Command> clone;
+        synchronized ( commandMap )
+        {
+            clone = new HashMap<>( commandMap );
```

Comment:
just do the unmodifiable here, why double clone?

Reply:
Collections.unmodifiableCollection(commandMap.entrySet()) is actually a minor bug in that it exposes the mutability of Map.Entry objects. This could be potentially used to mutate the underlying `commandMap` with `Entry#setValue`, bypassing synchronisation.

*unmodifiableCollection* and *unmodifiableMap* don't do any cloning. They're just wrappers.

Reply:
In that case we might want to deprecate the current method and add one which returns a Map where we can use unmodifiableMap.

Reply:
In my opinion there's no reason to do that. `Collections.unmodifiableMap( clone ).entrySet()` accomplishes the goal.